### PR TITLE
Improved hash managing

### PR DESCRIPTION
**Pull request by korgan00**
## Changes

Depends on https://github.com/Qiskit/saiba/pull/292
Jump to the section on page load.
Scrolling modifies the hash in the URL.
Scrolling highlights the section in the nav bar.

### CODE REVIEW

**Comment by abdonrd** [comment]
When I click a link, the scroll is under the toolbar (and no smooth):

https://user-images.githubusercontent.com/1007051/203178471-14b0bcb8-4b3f-4133-9e6c-4f9c82a3f5dd.mov

File: components/course-sidebar/Content.vue [hunk 1]
```diff
@@ -91,8 +100,44 @@ function isActiveUnit(unit: Unit | undefined) {
...
-  return `#${section.id}` === useRoute().hash;
+  return `#${section.id}` === hash.value;
 }
+
+let headingsY = [] as Ref<number>[];
+const { y } = useWindowScroll();
+
+watch(activeUnit, (unit) => {
+  // eslint-disable-next-line no-undef
+  nextTick(() => {
```
Comment (abdonrd) [inline components/course-sidebar/Content.vue:111]: If it's an auto-import from Vue, you can add it here:
https://github.com/Qiskit/saiba/blob/cacfd6dd96ce2707523252796ee5d266238246f8/packages/eslint-config/index.js#L22-L39

File: components/course-sidebar/Content.vue [hunk 2]
```diff
@@ -91,8 +100,44 @@ function isActiveUnit(unit: Unit | undefined) {
...
+    if (unit && unit.sections) {
+      headingsY =
+        unit.sections.map(({ id }) => {
+          const element = document.querySelector(`#${id}`) as MaybeElement;
+          return useElementBounding(element).y;
+        }) || [];
+      // eslint-disable-next-line no-undef
+      const router = useRouter();
```
Comment (abdonrd) [inline components/course-sidebar/Content.vue:119]: If it's an auto-import from Vue, you can add it here:
https://github.com/Qiskit/saiba/blob/cacfd6dd96ce2707523252796ee5d266238246f8/packages/eslint-config/index.js#L22-L39

File: components/course-sidebar/Content.vue [hunk 3]
```diff
@@ -65,6 +65,13 @@ import { useCourseStore } from '@/store/course/store';
 import { useUnitStore } from '@/store/unit/store';
 import { useUserStore } from '@/store/user/store';
 import { Section, Unit } from '@/types/unit';
+import {
+  MaybeElement,
+  useElementBounding,
+  useWindowScroll,
+  watchDebounced,
+} from '@vueuse/core';
+import { Ref } from 'vue';
```
Comment (abdonrd) [inline components/course-sidebar/Content.vue:74]:
```suggestion
import { type Ref } from 'vue';
```

Ot

File: app/router.options.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import type { RouterConfig } from '@nuxt/schema';
+// https://router.vuejs.org/api/interfaces/routeroptions.html
+export default <RouterConfig>{
+  scrollBehavior(to) {
+    if (to.hash) {
+      const toElement = document.querySelector(to.hash);
+      if (toElement) {
+        return {
+          top: window.scrollY - 64 + toElement.getBoundingClientRect().y,
```
Comment (abdonrd) [inline app/router.options.ts:9]: Where does this `64` come from?

Comment (korgan00) [reply app/router.options.ts:9]: Size of the top bar (3 rem) + 1 rem extra as padding. In our case 1rem = 16px

**Comment by korgan00** [comment]
> When I click a link, the scroll is hidden under the toolbar (and no smooth):
> 
>  Screen.Recording.2022-11-21.at.6.26.29.PM.mov

It is related to using `<a>` instead of `NuxtLink`/`BasicLink` and it doesn't raise the scroll behavior. It should be fixed now

File: components/course-sidebar/Content.vue (hunk 1 above)
Comment (korgan00) [reply components/course-sidebar/Content.vue:110]: I didn't know where it was moved :)

**Review by techtolentino — APPROVED** [review]
This looks great @korgan00 

👏🏼 

https://user-images.githubusercontent.com/6276074/203383063-7dbd422b-8fdc-45a7-bbce-c8b5101ce654.mov

File: app/router.options.ts (hunk 4 above)
Comment (korgan00) [reply app/router.options.ts]: Done, resolve the conversation if the new implementation is ok for you.

File: components/course-sidebar/Content.vue [hunk 5]
```diff
@@ -82,16 +89,59 @@ const { setActiveUnitIdBySlug } = unitStore;
...
+
+  headingsYPosition =
+    unit.sections.map(({ id }) => {
+      const element = document.querySelector(`#${id}`) as MaybeElement;
+      return useElementBounding(element).y;
+    }) || [];
+
+  const router = useRouter();
+  // Timeout to wait until page is fully rendered to prevent jumping.
+  setTimeout(() => router.replace({ hash: useRoute().hash }), 1000);
```
Comment (abdonrd) [inline components/course-sidebar/Content.vue:124]: Any other options? `setTimeout`s always scare me.

@Qiskit/qiskit-digital-devs thoughts?

**Review by abdonrd — COMMENTED** [review]
LGTM!

File: components/course-sidebar/Content.vue (hunk 5 above)
Comment (y4izus) [reply components/course-sidebar/Content.vue:124]: I agree with you @abdonrd , but Goyo already looked for other solutions and that is the best we have right now 🤷🏽‍♀️

**Review by y4izus — APPROVED** [review]
LGTM!! Thank you!

**Pull request merged**